source/common/common/assert.h

```diff
@@ -156,9 +156,19 @@ void resetEnvoyBugCountersForTest();
 // _ASSERT_VERBOSE, and this will call _ASSERT_VERBOSE,(__VA_ARGS__)
 #define ASSERT(...)                                                                                \
   EXPAND(_ASSERT_SELECTOR(__VA_ARGS__, _ASSERT_VERBOSE, _ASSERT_ORIGINAL)(__VA_ARGS__))
+
+// SLOW_ASSERTs should mimic ASSERTs in debug mode.
+#if !defined(NDEBUG) // debug build
+#define SLOW_ASSERT(...) ASSERT(__VA_ARGS__)
+#else // ENVOY_LOG_DEBUG_ASSERT_IN_RELEASE is defined
+// SLOW_ASSERTs in release mode should resolve to a non-implementation.
+#define SLOW_ASSERT _NULL_ASSERT_IMPL
```

Comment:
Would it make sense to split ENVOY_LOG_DEBUG_ASSERT_IN_RELEASE into 2 definitions:

ENVOY_LOG_FAST_DEBUG_ASSERT_IN_RELEASE
ENVOY_LOG_SLOW_DEBUG_ASSERT_IN_RELEASE

Defining ENVOY_LOG_DEBUG_ASSERT_IN_RELEASE would define the two new names that control the behavior of the various ASSERT flavors in release mode.

Many thanks for taking on the work to define these new assert types!

Reply:
Am I interpreting this correct:
ENVOY_LOG_FAST_DEBUG_ASSERT_IN_RELEASE: turns on ASSERTs, no-ops SLOW_ASSERT
ENVOY_LOG_SLOW_DEBUG_ASSERT_IN_RELEASE: turns on SLOW_ASSERT (? on ASSERT)

Why not have this so we keep existing behavior and get the fine tuning you suggest:
ENVOY_LOG_DEBUG_ASSERT_IN_RELEASE: turns on ASSERT, SLOW_ASSERT
ENVOY_LOG_FAST_DEBUG_ASSERT_IN_RELEASE: turns on ASSERT, no-ops SLOW_ASSERT

Reply:
The implementation is not bad with your suggestion.

Testing is a little bit hard because we need a new compilation job to do this since defines won't propagate to rebuilt `assert_lib`. I may need to create a new test-only `assert_with_fast_asserts_test_only_lib` target that has the define `ENVOY_LOG_FAST_DEBUG_ASSERT_IN_RELEASE` and create a new test with that target.

Reply:
Actually, I can just add a new line to just test `assert_test` with the ENVOY_LOG_FAST_DEBUG_ASSERT_IN_RELEASE. It won't add too much time to the CI job.
https://github.com/envoyproxy/envoy/blob/5e7eb111f4ac4af6bd8ec1be7169d6f1e982475c/ci/do_ci.sh#L316